### ISSUE TITLE
Missing validation of schema directive argument types

### ISSUE THREAD

**Issue by vwkd** [graphql/graphql-js#3912]
It seems there is no validation for schema directive argument types.

This makes the `kind` of the directive end up with the wrong type. Notably, if the type doesn't exist, it ends up as `EnumValue` as this seems to be the default type.

This is different from field argument types which correctly throw an error message if you pass the wrong type.

## Steps to reproduce

1. Run with deno, e.g. open `deno repl` and paste in, or run from file with `deno run file.ts`

```
import { buildSchema } from "npm:graphql@16.6.0";

const schemaSource = `
  type Query {
    # notice the missing quotes around the string!
    age: Int @foo(bar: FOOBAR)
  }

  directive @foo(
    bar: String!
  ) on FIELD_DEFINITION
`;

const schema = buildSchema(schemaSource);
// expected throw

const kind = schema.getQueryType()!.getFields()["age"].astNode!.directives![0].arguments![0].value.kind;

console.log(kind);
// EnumValue
```

## Expected result

The `buildSchema` function throws.

## Actual result

The `buildSchema` doesn't throw.

**Comment by benjie** [comment]
Agree it should be in section 3; but not sure whether it should be throughout or centralized. We define directive _definition_ validation here:

https://spec.graphql.org/draft/#sec-Type-System.Directives.Validation

For each of the possible schema directive locations (not runtime locations) we need to add to their validation rules that each of the directives applied pass the same or similar checks as in https://spec.graphql.org/draft/#sec-Validation.Directives and https://spec.graphql.org/draft/#sec-Values.

E.g. we already have "for each field of an Object type..."

https://github.com/graphql/graphql-spec/blob/2073bc888d92692cd465db47cc544109db9c4181/spec/Section%203%20--%20Type%20System.md#L907-L922

We could add "For each directive applied to an Object type..." and do so for each of the directive locations; but I think that would get a bit redundant.

Instead, we should maybe add another section immediately below https://spec.graphql.org/draft/#sec-Type-System.Directives.Validation that is something like "Application validation" that validates the applied directives across all of their locations.

**I think the main issue is we don't really have a "validate SDL" part of the spec**, and the _schema_ itself (in spec terms, not necessarily implementations) doesn't contain "applied directives" - the idea is that tooling would apply the directives by making changes to the schema, but the resulting schema would not "store" that these changes have happened. E.g. `fooConnection: [Foo] @connection` would have the effect of turning this into `fooConnection: FooConnection` and the `@connection` directive would evaporate. There are a few exceptions such as `@deprecated` where they get converted into something like `isDeprecated`/`deprecationReason` and then that can be converted back again when printing the schema SDL.

**Comment by captbaritone** [comment]
> the idea is that tooling would apply the directives by making changes to the schema, but the resulting schema would not "store" that these changes have happened

Yeah, I think that's a good explanation for why we find ourselves in this state.

That said, I think that we now have several examples of directives that are used to encode additional information that consumers/clients want to be able to see. `@semanticNonNull` and the directives used as part of the [composite schema spec](https://graphql.github.io/composite-schemas-spec/draft/#sec--lookup) both seem like they fit that bill to me.

(This next bit is veering out on a tangent a bit)
My understanding is that directives were intentionally added to GraphQL as an escape hatch (pressure release?) to allow users to extend the language. If that's their intended use, I think we need to consider this use case more clearly at the spec level. That would include adding this validation (so that clients don't need to validate directives as they read them) and also some support for reading schema directives via introspection.

**Comment by benjie** [comment]
We should note that there are two use cases for schema directives:

1. to change how a schema entity is formed or behaves (e.g. `foo: [Foo] @connection` to convert to connection, or `bar: Bar @requires(level: ADMIN)` for access control)
2. to add metadata to a schema entity (e.g. for use by tooling)

Currently GraphQL only caters to the former (with the nuance that `@specifiedBy` and `@deprecated` are _arguably_ metadata); for discussion around the latter check out graphql/graphql-spec#300 and related discussions.

---

To your point, @yaacovCR, I think that validating a schema SDL would be opt-in, so tooling that works in interesting ways (e.g. merging remote SDL) would be down to the tooling to validate or not the inputs. Nothing in the spec requires an SDL document, or any GraphQL document for that matter, to be valid on its own; it's dependent on how it's used - currently we only require that an _executable_ document be valid in order to execute it (and even then, there are caveats around when you can skip validation). You're right that it's essential we continue to reflect this nuance if/when writing up SDL validation into the spec.

**Comment by yaacovCR** [comment]
Track in #3222
